### PR TITLE
Add helm option enableEgressGateway

### PR DESCRIPTION
**Pull request by anfernee**
The option turns on the egress gateway support, which allows users to
specify Egress NAT policies to redirect and SNAT the egress traffic that
leaves the cluster. It's false by default.

Towards: #13575

Signed-off-by: Yongkun Gui <ygui@google.com>

### CODE REVIEW

**Review by kaworu — CHANGES_REQUESTED** [review]
Hello @anfernee and thanks for the PR!

Question: should the flag stay hidden if we're exposing it through Helm?

[This failure](https://github.com/cilium/cilium/pull/15777/checks?check_run_id=2384965259) seems legit to me, and the [helm-docs](https://github.com/cilium/cilium/pull/15777/checks?check_run_id=2384965652) reordered the `enableEgressGateway` in the README.md for some reason (`make -C install/kubernetes` should fix it).


/cc @joestringer as it seems you have context about https://github.com/cilium/cilium/pull/14723.

File: install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml [hunk 1]
```diff
@@ -104,6 +104,7 @@ rules:
   - ciliumlocalredirectpolicies
   - ciliumlocalredirectpolicies/status
   - ciliumlocalredirectpolicies/finalizers
+  - ciliumegressnatpolicies
```
Comment (aanm) [inline install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml:107]: We need `ciliumlocalredirectpolicies/status`. I'm not so sure about `ciliumlocalredirectpolicies/finalizers`. But we should probably put it to be on the safe side?

**Review by aanm — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml (hunk 1 above)
Comment (anfernee) [reply install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml:107]: ciliumegressnatpolicies doesn't have `status` field. but I guess there is no harm adding it though. will follow the pattern here.

**Comment by anfernee** [comment]
test-me-please

**Comment by anfernee** [comment]
test-gke retest-4.9 retest-4.19

**Review by pchaigno — APPROVED** [review]

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 2]
```diff
@@ -677,6 +677,10 @@ data:
   disable-cnp-status-updates: "true"
 {{- end }}
 
+{{- if .Values.enableEgressGateway }}
```
Comment (pchaigno) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: That's inconsistent with other names. We should probably go for `egressGateway.enabled=true` instead. Unless we're sure we won't ever need other subflags to `egressGateway` in which case just `egressGateway=true` is enough.

**Review by pchaigno — CHANGES_REQUESTED** [review]
Oops, almost missed something :-)

File: install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml (hunk 1 above)
Comment (aanm) [reply install/kubernetes/cilium/templates/cilium-agent-clusterrole.yaml:107]: @anfernee if that is the case then we should remove it from the CRD as it is currently defined in there https://github.com/cilium/cilium/blob/a783f7c55d34b647e751afe6fc7e892b1a809485/examples/crds/ciliumegressnatpolicies.yaml#L184-L185
https://github.com/cilium/cilium/blob/a783f7c55d34b647e751afe6fc7e892b1a809485/pkg/k8s/apis/cilium.io/v2alpha1/types.go#L27

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by anfernee** [comment]
@aanm good findings. I'll fix it.

**Comment by joestringer** [comment]
> Question: should the flag stay hidden if we're exposing it through Helm?

If we're now happy enough with the feature to put it in front of users and we don't intend on changing the flags any further, then we can un-hide the flag now as well:

```
diff --git a/daemon/cmd/daemon_main.go b/daemon/cmd/daemon_main.go
index 82af61b51ea3..2d22cd2c6655 100644
--- a/daemon/cmd/daemon_main.go
+++ b/daemon/cmd/daemon_main.go
@@ -667,7 +667,6 @@ func init() {

        flags.Bool(option.EnableEgressGateway, false, "Enable egress gateway")
        option.BindEnv(option.EnableEgressGateway)
-       flags.MarkHidden(option.EnableEgressGateway)

        flags.String(option.IPMasqAgentConfigPath, "/etc/config/ip-masq-agent", "ip-masq-agent configuration file path")
        option.BindEnv(option.IPMasqAgentConfigPath)
```

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
test-me-please

**Review by kaworu — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
The failure in the test-1.16-netnext was caused by jenkins rebasing the test code but the cilium code is not rebased in the image build process. The test had enabled tunneling mode in https://github.com/cilium/cilium/pull/15716 but this PR was not rebased against master.

**Comment by anfernee** [comment]
test-me-please

**Pull request merged**